flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/NullableSerializer.java

```diff
@@ -0,0 +1,221 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.java.typeutils.runtime;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeutils.CompatibilityResult;
+import org.apache.flink.api.common.typeutils.CompatibilityUtil;
+import org.apache.flink.api.common.typeutils.CompositeTypeSerializerConfigSnapshot;
+import org.apache.flink.api.common.typeutils.TypeDeserializerAdapter;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.TypeSerializerConfigSnapshot;
+import org.apache.flink.api.common.typeutils.UnloadableDummyTypeSerializer;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputSerializer;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.util.Preconditions;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * Serializer wrapper to add support of null value serialization.
+ *
+ * <p>If the target serializer does not support null values of its type,
+ * you can use this class to wrap this serializer.
+ * This is a generic treatment of null value serialization
+ * which comes with the cost of additional byte in the final serialized value.
+ * The {@code NullableSerializer} will intercept null value serialization case
+ * and prepend the target serialized value with a boolean flag marking whether it is null or not.
+ * <pre> {@code
+ * TypeSerializer<T> originalSerializer = ...;
+ * TypeSerializer<T> serializerWithNullValueSupport = NullableSerializer.wrap(originalSerializer);
+ * // or
+ * TypeSerializer<T> serializerWithNullValueSupport = NullableSerializer.wrapIfNullIsNotSupported(originalSerializer);
+ * }}</pre>
+ *
+ * @param <T> type to serialize
+ */
+public class NullableSerializer<T> extends TypeSerializer<T> {
+	private static final long serialVersionUID = 3335569358214720033L;
+
+	private final TypeSerializer<T> originalSerializer;
+
+	private NullableSerializer(TypeSerializer<T> originalSerializer) {
+		Preconditions.checkNotNull(originalSerializer, "The original serializer cannot be null");
+		this.originalSerializer = originalSerializer;
+	}
+
+	/**
+	 * This method tries to serialize null value with the {@code originalSerializer}
+	 * and wraps it in case of {@link NullPointerException}, otherwise it returns the {@code originalSerializer}.
+	 */
+	public static <T> TypeSerializer<T> wrapIfNullIsNotSupported(TypeSerializer<T> originalSerializer) {
+		return checkIfNullSupported(originalSerializer) ? originalSerializer : wrap(originalSerializer);
+	}
+
+	private static <T> boolean checkIfNullSupported(TypeSerializer<T> originalSerializer) {
+		try {
+			originalSerializer.serialize(null, new DataOutputSerializer(1));
+			Preconditions.checkArgument(originalSerializer.copy(null) == null);
+		} catch (NullPointerException | IOException e) {
```

Comment:
This triggered another idea, maybe we should also test for this in the `SerializerTestBase`, for example that `null` either works or throws an "acceptable" exception. In case that it works, we could also check other properties, e.g. that if the serializer declares a fixed size also the serialized bytes for `null` have the same size.